### DISCUSSION TITLE
gRPC support

### DISCUSSION THREAD

**Discussion by Clement-Jean** [curl/curl#11239] [Ideas]
Hey everyone. I'm interested in starting discussion about adding support to gRPC. Since cURL already support HTTP/2, I guess the main work would be parsing Protobuf (not sure though).

I am aware that there is a project called grpcurl but I believe adding native support in cURL would be bringing more adoptions to both gRPC and cURL.

I'm interested in getting opinion from you guys. Would that make sense? I would be happy totry wotking on that with other contributors.

**Comment by bagder** [comment]
I don't know much about gRPC, I have never seen nor tried grpcurl and I don't know the intended use cases for this. So let me ask:

Is the primary use case you see for this a way to specify the data to send to the gRPC server (text2binary) and then have the reverse functionality (binary2text) for the response, so that you can convert the response blob to something readable? Or is there more to it?

**Comment by Clement-Jean** [comment]
The MVP kind of feature is what you describe. gRPC is effectively Protobuf over HTTP/2. We need to be able to transform either JSON or Protobuf Text Format into binary (following Protobuf serialization rules) and do the opposite.

However, there is definitely more to it. For example:

- gRPC supports unary, server stream, client stream, and bidi stream. This would probably need some way of sending and displaying multiple requests/responses. Currently grpcurl use interactive terminals to do that. An example of this is:

```shell
grpcurl -d @ 0.0.0.0:50051 $SERVICE.$METHOD <<EOF
{ "description": "new description" }
{ "description": "new description" }
EOF
```

we use -d for specifying data and in this case we use @ to say that we send data through interactive terminal. And the <<EOF syntax is pretty standard.

- Notice that in the previous example we send some JSON as data. This data will be matched against a Protobuf schema and if it doesn't comply to the schema (e.g unknown field, etc..), the command will fail (without the -allow-unknown-fields flag).
- After that there is something called server reflection where the server exposes information about the services, methods and message that are register on it. This is nice for debugging and manual testing. An example of that is:

```shell
$ grpcurl 0.0.0.0:50051 describe [$SERVICE | $METHOD | $MESSAGE]
$ grpcurl 0.0.0.0:50051 list
```

where the first one give more info on the given parameter and the second one list the services register on the server.

- Protobuf also has a [text format](https://protobuf.dev/reference/protobuf/textformat-spec) which looks like JSON but is a little bit less verbose. I don't think grpcurl supports it but that could also be useful for test data in files.

Obviously there way more features that could be added but these are the ones that are majorly used outside of must have features like TLS, headers, ...

**Comment by bagder** [comment]
Since gRPC is a data protocol over HTTP/2 without any URL syntax of its own, I don't think the support belongs in libcurl.

As it is mostly a matter of converting from text to binary (and back) and send the data over h2, it seems to me like it should be perfectly doable to add support for gRPC in the curl tool.

One of the largest work items for implementing something like this is to make sure that we have test cases to verify good and bad uses to the extent that we can be fairly sure things are working properly.

Maybe a first work item would be to work out an idea and model for how to use the command tool to do gRPC. How would  the curl command lines look for the N most common use cases?

**Comment by Clement-Jean** [comment]
I was thinking of something like the following.

Because we do not have GET, POST, ... in gRPC, we could replace the values of -X by something like: UNARY, SSTREAM (server streaming), CSTREAM (client streaming), BSTREAM (Bidirectional streaming).

On top of that we will need to send some headers for gRPC to understand our requests. We need `Content-Type: application/grpc` and something like `Method: /$Package.$Service/$Method` or `/$Service/$Method` (if no package provided) to tell which endpoint we want to reach.

For the data, we could use -d and pass some JSON data to it. We could have a simple object for UNARY and SSTREAM and an array of objects for streams. This would look like this:

```shell
curl -H "Content-Type: application/grpc" -H "Method: /$Package.$Service/$Method" -X UNARY -d "{name: 'bagder'}" 0.0.0.0:50051
curl -H ... -X SSTREAM -d "{name: 'bagder'}" 0.0.0.0:50051
curl -H ... -X CSTREAM -d "[{name: 'bagder'}, {name: 'clement'}]" 0.0.0.0:50051
curl -H ... -X SSTREAM -d "[{name: 'bagder'}, {name: 'clement'}]" 0.0.0.0:50051
```

Finally, I believe the -X could be omitted when server reflection is enabled.

**Reply by bagder** [reply]
> Because we do not have GET, POST, ... in gRPC

but since gRPC is done over HTTP/2, we still have them involved don't we? They're just in the layer outside of gRPC. What HTTP methods are gRPC using otherwise?

> we could replace the values of -X by something like: UNARY, SSTREAM (server streaming), CSTREAM (client streaming), BSTREAM (Bidirectional streaming).

To me, it feels like this is better suited for a new option. An option that then also knows about the available options and will not accept any non-supported version. But maybe I'm not seeing the full picture yet.

> we will need to send some headers for gRPC to understand our requests

I think we can be inspired by how `--json` works: when used it will set a few headers. A `--grpc` option could work similarly. We need a way to tell the tool that it should speak gRPC for this transfer. Perhaps it also sets the stream type?

> we could use -d and pass some JSON data to it

this then requires JSON parsing ability, which curl currently does not have

> curl -H ... -X SSTREAM -d "[{name: 'bagder'}, {name: 'clement'}]" 0.0.0.0:50051

In this case, is that "0.0.0.0:50051" a shortcut for saying you send this to the gRPC server on port 50051 on locahost or is it something else? (50051 seems like such an peculiar port number to use in an example)

Do gRPC servers not act on a path on HTTP ? Can it not handle HTTP authentication or HTTP redirects?

**Comment by Clement-Jean** [comment]
Right. What I meant was that GET, POST, etc.. are not gRPC primitives. They are encapsulated by it.

> To me, it feels like this is better suited for a new option.

This sounds like a good idea. I don't know much about the internals of curl but if we can add a new dedicated option that would be nice.

> A --grpc option could work similarly

This would also be nice. We wouldn't have the manually set the application/grpc Content-Type header. The use of this flag would add it automatically.

> Perhaps it also sets the stream type?

I'm not really sure how the `--json` flag works (going to check out) but for the stream type what do you have in mind?

> this then requires JSON parsing ability, which curl currently does not have

Either we go with JSON or we would have to chose another format. I can see two other ways:

- A binary file. The problem is that this involves a first step for creating the binary file with code or with the protobuf compiler. I believe this is not user-friendly.
- Using the Protobuf [Text Format's message list](https://protobuf.dev/reference/protobuf/textformat-spec/#fields). However I believe this is very similar to JSON and thus why not supporting JSON directly.

Now the question is: should we add support to that in curl or should we rely on other library?

> In this case, is that "0.0.0.0:50051" ...

50051 is a port often used in gRPC. This is pretty much because most tutorials will use it but there is nothing tying gRPC application to this port.

> Do gRPC servers not act on a path on HTTP ? Can it not handle HTTP authentication or HTTP redirects?

I'm not entirely sure about this for the lower level, but at the user level we never work with path, this is the point of RPCs. authentication is feasible but through other means (interceptors, ...), and I'm not aware of any redirects happening in gRPC.

**Reply by bagder** [reply]
> Right. What I meant was that GET, POST, etc.. are not gRPC primitives. They are encapsulated by it.

Right. From my brief research, it seems POST is always used.

> for the stream type what do you have in mind?

Based on your earlier examples I got the impression a stream type needs to be set? So maybe `--grpc [type]` would work? Tells curl to use gRPC and sets the type, at once.

> 50051 is a port often used in gRPC

Ok. BTW, are those services then typically used over TLS or not?

**Reply by Clement-Jean** [reply]
> So maybe --grpc [type] would work?

This looks good to me. And it's pretty short to type.

> BTW, are those services then typically used over TLS or not?

Yes. Actually, without explicitly telling them to use plaintext, gRPC will assume that it's using TLS, mTLS, or ATLS. See  [https://grpc.io/docs/guides/auth/](https://grpc.io/docs/guides/auth/) for learning more about all the auth possible in gRPC.

**Reply by bagder** [reply]
> without explicitly telling them to use plaintext, gRPC will assume that it's using TLS

It makes it really odd that you connect to it with `0.0.0.0:50051` because then you can't verify the server certificate. I suppose it might tell us that grpcurl isn't too particular about that... With curl, you would be encouraged to use the full `https://servername:50051`.

**Reply by Clement-Jean** [reply]
Alright. I forgot to mention that we do not add the URL scheme in gRPC Golang URLs. it's in format ${IPV4/IPV6}:${PORT} all the other configuration happens either through options, or are handled by an external service (e.g. load balancer).

For other implementations which are not pure (not implemented in the in the same language but are wrappers around the C++ lib), we have the following [URL schemes](https://grpc.github.io/grpc/cpp/md_doc_naming.html).

I guess we should focus on the C++ lib standard because there are a lot of wrappers around it. That would cover most of the [implementations](https://grpc.io/docs/languages/) (all except Go and C#) and then we can tweak for making it work with Go or C# (not sure how C# handles URLs right now)

**Reply by bagder** [reply]
> we do not add the URL scheme in gRPC Golang URLs

Strictly speaking, a URL always have a scheme so if there is no scheme it is not a URL.

curl allows "scheme-less" URLs as a convenience thing but then it guesses which scheme that was meant and it will default to HTTP, meaning no TLS and that will fail miserably if the server uses TLS. Also: a URL without a host name cannot successfully verify the server name in the server certificate.

I don't think we can easily avoid using `https://` URLs when doing gRPC with curl. After all, gRPC *is* mostly HTTPS so I don't think that seems terribly bad.

> all the other configuration happens either through options, or are handled by an external service (e.g. load balancer).

When you invoke a curl command line, none of those other options are available.

> we have the following URL schemes

Yuck. They are going to cause friction if we try to use those.

> I guess we should focus on the C++ lib standard

There are different standards based on the used language? What kind of horror is this?

**Reply by Clement-Jean** [reply]
> it will default to HTTP

Could we default to HTTPS and fail miserably if the server doesn't use TLS? gRPC tries to provide secure connections by default.

> a URL without a host name cannot successfully verify the server name in the server certificate

Right. I was showing an example on localhost (in dev). For those we generally set an authority header and self-signed certificates. In prod we would have either a hostname or a authority header set with CA certificates.

> none of those other options are available

yes, we would probably need to add some of them if possible. An example (in grpcurl) of that is `-plaintext` and `-insecure` which allows non TLS connection and skip certificate verification respectively.

> They are going to cause friction if we try to use those.

Hmmm. Then we could use the Go Implementation approach but we would have to add more options. For example, grpcurl provides a `-unix` flag for explaining that the server address is a unix domain socket.

> There are different standards based on the used language?

My guess is that this is due to the networking code in the different standard libraries. But once again most of the implementations are wrappers around the C++ implementation which makes them work with the URL schemes listed earlier.

**Reply by bagder** [reply]
> Could we default to HTTPS and fail miserably if the server doesn't use TLS?

We could probably consider that, sure, but it seems a little premature to decide those teeny details already.

> Right. I was showing an example on localhost (in dev). For those we generally set an authority header and self-signed certificates.

Self-signed is still signed, there's no difference in verification method for curl for that. Just a matter of that CA store you would use.

We also have `--insecure / -k` in curl but we do not like to promote the use of this since it tends to stick and creep into production.

**Reply by Clement-Jean** [reply]
So it seems that we could do it right? 

Should we try to list the options and the usage somehow? Maybe like an output of --help

Is there any process that we need to go through in order to consider implementing this? a document, a vote, or something like this.

Is there anything your not convinced of? The usefulness, the feasibility, etc...

**Reply by bagder** [reply]
I think all this is doable and I don't think we have identified any show-stopper.

Laying out some ideas in terms of how dedicated command line options are envisioned to work is good and useful, as it will also offer other interested people a chance to ponder and contribute feedback already before there is code to test. There is no doubt in my mind that these things might need a little experimenting and offering users to "get a feel" for them before we know for sure how everything should be done.

There is no process. I think gRPC support already falls into the realm of what could be acceptable functionality for curl. **20%** of curl users answered in the 2023 user survey that they would like to see "gRPC support".

We normally ship brand new features/protocols as "experimental" first for some time, to let everyone try it out first and allow us to change things before we carve the rules in stone and they are never to be changed again. We should do that for gRPC as well.

All this said: there is a lot of work needed before we are there:

- code and sensibly working command line options (for old and new gPRC users, for new and old curl users, on multiple platforms)
- documentation - everyone needs to be able to learn how to use this
- tests - an important part as we cannot know that everything works solidly and safely until we have a decent set of dedicated tests for the functionality

**Reply by Clement-Jean** [reply]
Is there some kind of tutorial, template, or project that I could take a look at in order to learn how to support new protocols?

**Reply by bagder** [reply]
Not really, no.

I think you should start with not thinking about it as a new protocol. Protocols in curl are usually the URL scheme and since gRPC is still HTTP(S), we do not think of it as a new *protocol*. It is more of a *feature* for HTTP(S).

As an absolutely first shot you probably want to try to just store a binary gRPC message in a file on disk and then send that to a gRPC server and see how that works:

~~~
curl --data-binary @grpcfile -H "Content-Type: application/grpc" https://grpc.example.com:50051/ -o grpcresponse
~~~
(I'm guessing, it might need a little more tweaking to be accepted.)

Then consider the next steps to take to convert that into a more convenient command line.

**Comment by Clement-Jean** [comment]
Ok. Using a unary API is indeed feasible. I will try my best to document what I learned so far.

Let's start with the following proto file (greet.proto):

```proto
syntax = "proto3";

package greet;

message GreetRequest {
  string first_name = 1;
}

message GreetResponse {
  string result = 1;
}

service GreetService {
  rpc Greet(GreetRequest) returns (GreetResponse);
};
```

It is not that important what the RPC endpoint is doing but for this example I simply made it return "Hello " + the first_name sent in `GreetRequest`.

From that we can create a binary representation of the `GreetRequest` message. To make it simple we can use the Protobuf Text Format and generate binary with the protoc --encode flag. `request.txt` will look like this:

```
first_name: "Clement"
```

and then we execute the following command to get the binary:

```shell
(cat request.txt | protoc --encode=greet.GreetRequest greet.proto) > payload.bin
```

Now, using the payload.bin is not enough to make a valid request. After checking some WireShark logs, I realized that we need to prepend the Protobuf binary with a byte saying wether or not the binary is compressed and the equivalent of Content-Length header (which seems to be skipped by gRPC). So let's start by serializing 0 (not compressed) in a `request.bin` file:

```shell
printf "0: %.2x" 0 | xxd -r -g0 >> request.bin
```

Then we append 4 bytes equal to 9 (the length of Clement + 2 bytes for Protobuf metadata) in `tmp`:

```shell
printf "0: %.8x" 9 | xxd -r -g0 >> request.bin
```

And finally, we can add the Protobuf payload after that:

```shell
cat payload.bin >> request.bin
```
`request.bin` is now usable for sending a request. For now, I only tried on localhost with and without TLS. Without TLS, we can execute the following command:

```shell
curl --http2-prior-knowledge --insecure --data-binary @request.bin --header "Content-Type: application/grpc" http://0.0.0.0:50051/greet.GreetService/Greet --output response.bin
```

and with TLS, we can use (insecure only because I'm on localhost):

```shell
curl --http2 --cacert ssl/ca.crt --tlsv1.2 --tls-max 1.2 --insecure --data-binary @request.bin --header "Content-Type: application/grpc" https://0.0.0.0:50051/greet.GreetService/Greet --output response.bin
```

We now have the `response.bin`. Once again, there is some wrapping happening. We can skip the first 5 bytes (1 for compression, 4 for Content-Length) and decode the payload into `GreetResponse`:

```shell
tail -c +6 response.bin | protoc --decode=greet.GreetResponse greet.proto
result: "Hello Clement"
```

I need to admit this is quite hard for now since I did it by hand in bash. However, this can definitely by solved in code. I will try the other kind of Endpoints (server, client and bidi streaming). Hope this helps

**Comment by Clement-Jean** [comment]
I was able to make an MVP by using libcurl: https://github.com/Clement-Jean/grpc-curl. The integration and API design still needs to be discussed.

**Reply by Clement-Jean** [reply]
Making calls using server reflection is also feasible. The overall idea seems promising. I'll reread the previous messages in this thread and come up with a document to integrate with cURL.

**Comment by Clement-Jean** [comment]
# Proposal

This proposal describes the API for the gRPC protocol in cURL. We will first talk about the `--grpc` flag, then we will describe how the commands would look like for the most common uses cases (unary, server streaming, client streaming, bidi streaming), and then we will talk about the need to introduce some flags.

## --grpc

The `--grpc` flag would be introduced, similarly to other protocols like `--http1.1`, `--http2`, ... This is needed because gRPC has a custom HTTP/2 frame that looks like the following:

- compressed+frame_type (1 byte)
- request/response size (4 bytes)
- the protobuf payload  (n bytes)

With the flag introduced we would also be able to implicitly set the `Content-Type` to `application/grpc+proto` and know what frame parsing should be used.

## Background

For the remaining examples, we will use this fictional gRPC API:

- `GreetOnce` is a unary RPC which that takes a `GreetOnceRequest` and returns a `GreetOnceResponse`.
  - `GreetOnceRequest` has a field called `name` of type string.
  - `GreetOnceResponse` has a field called `greeting` of type string.
  - `GreetOnceResponse.greeting` is the concatenation of `"Hello "` with the `GreetOnceRequest.name`.
  - e.g. `{ "name": "badger" }` -> `{ "greeting": "Hello badger" }`

- `GreetManyTimes` is a server streaming RPC that takes a `GreetManyTimesRequest` and returns one or more `GreetManyTimesResponse`.
  - `GreetManyTimesRequest` has a field called `name` of type string and a field called `n` of type uint64.
  - `GreetManyTimesResponse` has a field called `greeting` of type string.
  - `GreetManyTimesResponse.greeting` is the concatenation of `"Hello "` with the `GreetManyTimesRequest.name`.
  - `GreetManyTimes` will return `GreetManyTimesRequest.n` responses.
  - e.g. 
    `{ "name": "badger", "n": 5 }` -> `{ "greeting": "Hello badger" }`
      -> `{ "greeting": "Hello badger" }`
      -> `{ "greeting": "Hello badger" }`
      -> `{ "greeting": "Hello badger" }`
      -> `{ "greeting": "Hello badger" }`

- `GreetJointly` is client streaming RPC that takes one or more `GreetJointlyRequest` and returns a `GreetJointlyResponse`.
  - `GreetJointlyRequest` has a field called `name` of type string and a field called `separator` of type string.
  - `GreetJointlyResponse` has a field called `greeting` of type string.
  - `GreetJointlyResponse.greeting` is the concatenation of `"Hello "` with all the `GreetJointlyRequest.name` interleaved with `GreetJointlyRequest.separator`.
  - e.g. 
    `{ "name": "badger", "separator": " and " }`
    `{ "name": "clement" }`
     -> `{ "greeting": "Hello badger and clement" }`

- `GreetSeparately` is a bidirectional streaming RPC that takes one or more `GreetSeparatelyRequest` and returns one or more `GreetSeparatelyResponse`.
  - `GreetSeparatelyRequest` has a field called `name` of type string.
  - `GreetSeparatelyResponse` has a field called `greeting` of type string.
  - `GreetSeparatelyResponse.greeting` is the concatenation of `"Hello "` with the `GreetSeparatelyRequest.name`.
  - e.g.
    `{ "name": "badger" }` -> `{ "greeting": "Hello badger" }`
    `{ "name": "clement" }` -> `{ "greeting": "Hello clement" }`

All these RPCs are under the `greet` package and in the `GreetService` service. So, for example, the full path to access `GreetOnce` is `greet.GreetService/GreetOnce`.

### URL

The URL to a gRPC endpoint will be of format: `[http|https|unix]://ip_or_domain_name/(pkg.)service/method`.

- `http` for plaintext communication.
- `https` for encrypted communication.
- `unix` for communication over unix socket.
- The package (`pkg`) is optional in protobuf.

### Reflection

gRPC provides a feature called reflection that let the clients get information about the API itself. You can:

- list all the services registered in the gRPC server.
- get metadata about a service to show all the methods.
- get metadata about a method to know which input/output type it needs.
- get metadata about a message to show its fields.
- check in which proto file a message/service is defined.

This means that, with reflection enabled, a client like `curl` can get all the metadata it needs about an RPC without needing to read a schema in a proto file. For example, with reflection, we can know that `GreetOnce` is a unary RPC accepting a `GreetOneRequest` and returning a `GreetOnceResponse`, without needing access to the Protobuf schema.

For the sake of clarity, in the following examples, we will assume that reflection is enabled on the server. Later on, we will see what we can do when reflection is disabled.

## Unary RPC

```shell
$ curl --grpc --data '{ "name": "badger" }' http://localhost:50051/greet.GreetService/GreetOnce
{
  "greeting": "Hello badger"
}
```

An unary RPC is very similar to normal HTTP requests. We simply pass the `data` we want to send as request.

## Server Streaming

```shell
$ curl --grpc --data '{ "name": "badger", "n": 5 }' http://localhost:50051/greet.GreetService/GreetManyTimes
{
  "greeting": "Hello badger"
}
{
  "greeting": "Hello badger"
}
{
  "greeting": "Hello badger"
}
{
  "greeting": "Hello badger"
}
{
  "greeting": "Hello badger"
}
```

Server streaming is also very similar to normal HTTP requests. The main difference will be that we can get more than one response.

## Client Streaming

```shell
$ curl --grpc --data @- http://localhost:50051/greet.GreetService/GreetJointly
{ "name": "badger", "separator": " and " }
{ "name": "clement" }
[CTRL+d]
{
  "greeting": "Hello badger and clement"
}
```

or

```shell
$ curl --grpc --data @- http://localhost:50051/greet.GreetService/GreetJointly <<EOF
{ "name": "badger", "separator": " and " }
{ "name": "clement" }
EOF
{
  "greeting": "Hello badger and clement"
}
```

Client streaming needs input from the CLI. We can get that with the `data` set to `@-` and either use heredocs or a `getline` loop. Finally, on EOF, the server will send a response back.

## Bidirectional Streaming

```shell
$ curl --grpc --data @- http://localhost:50051/greet.GreetService/GreetJointly
{ "name": "badger" }
{
  "greeting": "Hello badger"
}
{ "name": "clement" }
{
  "greeting": "Hello clement"
}
```

or

```shell
$ curl --grpc --data @- http://localhost:50051/greet.GreetService/GreetJointly <<EOF
{ "name": "badger" }
{ "name": "clement" }
EOF
{
  "greeting": "Hello badger"
}
{
  "greeting": "Hello clement"
}
```

Bidirectional streaming also needs input from the CLI. However, it also needs to be able to read responses from the server in real time. This is because a bidirectional streaming API can return a response at any time. Here we show an example of an API return 1 response per request but you could think about an API returning a response every 5 requests or every 10 minutes after the initial request.

This means that we can implement this with a `getline` loop just like the client streaming, but heredocs would act as a kind of `replay`.

## Other Flags

Here are the other, already existing, flags that can be used:

- `--header` to set RPC headers.
- `--cacert` to set the path to a CA certificate to be used for encrypted communication.
- `--compressed` to tell the server to return a compressed response.
- `--unix-socket` to talk over a given unix socket.

Some flags will need to be added:

- `--proto <path>`/`-p <path>` to specify the proto file(s) in which the messages/services are defined.
  - this flag will disable the rather expensive (requires another API call) check for reflection.
  - e.g. `curl --grpc --data '{ "name": "badger" } --proto greet.proto http://localhost:50051/greet.Greetservice/GreetOnce'` tells us that the `greet.GreetService/GreetOnce` method is defined in `greet.proto`.

- `--proto_path <path>`/`-I <path>` to specify the root path(s) for Protobuf imports.
  - e.g. `curl --grpc --data '{ "name": "badger" } --proto greet.proto --proto_path proto http://localhost:50051/greet.Greetservice/GreetOnce'` tells us that the import defined in `greet.proto` are situated in the proto directory.

And other flags can be considered:

- `-X LIST` to list all the services registered in the server with reflection.

  ```shell
  $ curl --grpc -X LIST http://localhost:50051
  greet.GreetService
  grpc.reflection.v1.ServerReflection
  grpc.reflection.v1alpha.ServerReflection
  ```

- `-X DESCRIBE <method|message|service>` to get the metadata of a given method, message, or service (on server with reflection enabled).
  - for a message:

    ```proto
    $ curl --grpc -X DESCRIBE http://localhost:50051/greet.GreetOnceRequest
    message GreetOnceRequest {
      string name = 1;
    }
    ```

  - for a method:

    ```proto
    $ curl --grpc -X DESCRIBE http://localhost:50051/greet.GreetService/GreetOnce
    rpc GreetOnce ( .greet.GreetOnceRequest ) returns ( .greet.GreetOnceResponse );
    ```

  - for a service:

    ```proto
    $ curl --grpc -X DESCRIBE http://localhost:50051/greet.GreetService
    service GreetService {
      rpc GreetOnce ( .greet.GreetOnceRequest ) returns ( .greet.GreetOnceResponse );
      rpc GreetManyTimes ( .greet.GreetManyTimesRequest ) returns ( stream .greet.GreetManyTimesResponse );
      rpc GreetJointly ( stream .greet.GreetJointlyRequest ) returns ( .greet.GreetJointlyResponse );
      rpc GreetSeparately ( stream .greet.GreetSeparatelyRequest ) returns ( stream .greet.GreetSeparatelyResponse );
    }
   ```

## Other Considerations

- It would be easier to implement the gRPC support in C++. But we could do some research to check if the `μpb` (or upb) official C library could be used.

- The easiest way to implement this proposal is to depend on`libprotobuf`. It provides:
  - Encoding of `google::protobuf::Message` to bytes.
  - Decoding of bytes to `google::protobuf::Message`.
  - A JSON to message utility (`JsonStringToMessage`)
  - A Message to JSON utility (`MessageToJsonString`)
  - A way to create message instances from their name (`DynamicMessageFactory`).

- We can implement reflection checking without having to depend on a gRPC library. We would simply copy over the Protobuf schema, and compile it with `protoc`.

- To provide a better developer experience, we will need to detect if the Protobuf compiler (`protoc`) is installed on the system. This is because, Protobuf comes with a bunch of Well-Known Types and we do not want users to have to specify the `--proto_path` for these types. However, we need to consider the problem of having multiple `protoc` installed on the system (we can either have a environment variable or always select the newest version) and the different ways of detecting that `protoc` is installed depending on the OS (this is already implemented in `protoc` itself).
